docs/spec.md

```diff
@@ -200,3 +200,151 @@ $ cat .gitattributes
 ```
 
 Use the `git lfs track` command to view and add to `.gitattributes`.
+
+## Extending LFS
+
+Teams who use Git LFS often have custom requirements for how the pointer files and 
+blobs should be handled.  Some examples of extensions that could be built:
+
+* Compress large files on clean, uncompress them on smudge/fetch
+* Encrypt files on clean, decrypt on smudge/fetch
+* Scan files on clean to make sure they don't contain sensitive information
+
+The basic extensibilty model is that LFS extensions must be registered explicitly, and
+they will be invoked on clean and smudge to manipulate the contents of the files as 
+needed.  On clean, LFS itself ensures that the pointer file is updated with all the 
+information needed to be able to smudge correctly, and the extensions never write to the 
+pointer file.
+
+Note that LFS is currently transitioning away from using the Git smudge filter, in favor 
+of smudging all files using "git-lfs fetch" post checkout.  However, that detail should 
+be transparent to extensions, since they are still invoked on a per-file basis.
+
+### Registration
+
+To register an LFS extension, it must be added to the Git config.  Each extension needs 
+to define:
+
+* Its unique name.  This will be used as part of the key in the pointer file.
+* The command to run on clean
+* The command to run on smudge/fetch
+* The priority of the extension, which must be a unique, positive integer
+
+The sequence "%f" in the clean and smudge commands will be replaced by the filename being 
+processed.
+
+Here's an example extension registration in the Git config:
+
+```
+[lfs-extension "foo"]
+  clean = git-lfs-foo clean %f
+  smudge = git-lfs-foo smudge %f
+  priority = 1
+[lfs-extension "bar"]
+  clean = git-lfs-bar clean %f
+  smudge = git-lfs-bar smudge %f
+  priority = 2
```

Comment:
I'm making a basic assumption here that since we're dealing with large files, the overhead of reading a file twice is higher than the overhead of spinning up a process to execute an extension.  Would you agree with that assumption?

If we do as you suggest, it would require the extension to have to read the file off disk, even though Git or LFS have already read that same file.  Streaming the file contents over STDIN avoids that double I/O, but it does add process overhead.

Longer term, if we see that process overhead is too high, we can make a more meaningful optimization, and allow extensions to run in-proc.  This will be especially helpful when using "git lfs fetch" since there is a long-running LFS process that can then make multiple calls to an extension API.  The downside is it places more constraints on extension developers.

Thoughts?